auction-server/src/simulator/main.rs

```diff
@@ -0,0 +1,191 @@
+use {
+    litesvm::{
+        types::{
+            FailedTransactionMetadata,
+            SimulatedTransactionInfo,
+        },
+        LiteSVM,
+    },
+    solana_client::{
+        rpc_config::RpcSendTransactionConfig,
+        rpc_response::{
+            Response,
+            RpcResult,
+        },
+    },
+    solana_rpc_client::nonblocking::rpc_client::RpcClient,
+    solana_sdk::{
+        account::{
+            Account,
+            ReadableAccount,
+        },
+        account_utils::StateMut,
+        bpf_loader_upgradeable::UpgradeableLoaderState,
+        commitment_config::CommitmentConfig,
+        pubkey::Pubkey,
+        signature::Signature,
+        transaction::VersionedTransaction,
+    },
+    std::{
+        collections::HashMap,
+        time::Instant,
+    },
+    tokio::sync::RwLock,
+};
+
+pub struct Simulator {
+    sender:      RpcClient,
+    receiver:    RpcClient,
+    pending_txs: RwLock<Vec<(VersionedTransaction, Instant)>>,
+}
+
+struct AccountsConfig {
+    accounts: HashMap<Pubkey, Account>,
+    programs: HashMap<Pubkey, Account>,
+}
+
+impl AccountsConfig {
+    fn new() -> Self {
+        Self {
+            accounts: Default::default(),
+            programs: Default::default(),
+        }
+    }
+
+    fn apply(&self, svm: &mut LiteSVM) {
+        for (key, account) in self.accounts.iter() {
+            svm.set_account(*key, account.clone()).unwrap();
+        }
+        for (key, account) in self.programs.iter() {
+            svm.add_program(
+                *key,
+                &(account.data()[UpgradeableLoaderState::size_of_programdata_metadata()..]),
+            );
+        }
+    }
+}
+
+impl Simulator {
+    pub fn new(sender: RpcClient, receiver: RpcClient) -> Self {
+        Self {
+            sender,
+            receiver,
+            pending_txs: Default::default(),
+        }
+    }
+
+    pub async fn fetch_pending_and_remove_old_txs(&self) -> Vec<VersionedTransaction> {
+        let now = Instant::now();
+        let mut pending_txs = self.pending_txs.write().await;
+        pending_txs.retain(|(_, time)| now.duration_since(*time).as_secs() < 5);
+        //TODO: remove pending txs when they are processed
+        pending_txs.iter().map(|(tx, _)| tx.clone()).collect()
+    }
+
+    pub async fn send_transaction(
+        &self,
+        tx: &VersionedTransaction,
+    ) -> solana_client::client_error::Result<Signature> {
+        let now = Instant::now();
+        self.pending_txs.write().await.push((tx.clone(), now));
+        self.sender
```

Comment:
why not simulate the tx here optimistically? you're anyways going to simulate it later, so why not do so here and poll regularly until the tx either hits `processed` or doesn't show up on-chain after some buffer time (in which case you revert to the RPC's view of state)

Reply:
didn't understand what you mean. We simulate it just before calling this function, why should we simulate it again?